tests/topotests/multicast_features/test_multicast_features.py

```diff
@@ -0,0 +1,349 @@
+    r1_r2_link_address = out["r1-eth0"]["ipAddresses"][1]["address"].split('/')[0]
+    out = tgen.gears["r1"].vtysh_cmd("show interface r1-eth1 json", True)
+    r1_r3_link_address = out["r1-eth1"]["ipAddresses"][1]["address"].split('/')[0]
+    out = tgen.gears["r2"].vtysh_cmd("show interface r2-eth0 json", True)
+    r2_link_address = out["r2-eth0"]["ipAddresses"][1]["address"].split('/')[0]
+    out = tgen.gears["r3"].vtysh_cmd("show interface r3-eth0 json", True)
+    r3_link_address = out["r3-eth0"]["ipAddresses"][1]["address"].split('/')[0]
+
+    expect_pim_peer("r1", "ipv6", "r1-eth0", r2_link_address)
+    expect_pim_peer("r2", "ipv6", "r2-eth0", r1_r2_link_address)
+    expect_pim_peer("r1", "ipv6", "r1-eth1", r3_link_address)
+
+
+def test_igmp_group_limit():
+    "Test IGMP group limits."
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    tgen.gears["r1"].vtysh_cmd("""
+        configure terminal
+        interface r1-eth2
+         ip igmp max-groups 4
+    """)
+    app_helper.run("h1", ["224.0.100.1", "h1-eth0"])
+    app_helper.run("h1", ["224.0.100.2", "h1-eth0"])
+    app_helper.run("h1", ["224.0.100.3", "h1-eth0"])
+    app_helper.run("h1", ["224.0.100.4", "h1-eth0"])
+    app_helper.run("h1", ["224.0.100.5", "h1-eth0"])
+    app_helper.run("h1", ["224.0.100.6", "h1-eth0"])
+
+    topotest.sleep(10, "Waiting to learn IGMP groups")
+
+    igmp_groups = tgen.gears["r1"].vtysh_cmd("show ip igmp groups json", isjson=True)
+    assert len(igmp_groups["r1-eth2"]["groups"]) == 4
+
+    # Cleanup
+    app_helper.stop_host("h1")
+    tgen.gears["r1"].vtysh_cmd("""
+        configure terminal
+        interface r1-eth2
+         no ip igmp max-groups 4
+        exit
+        clear ip igmp interfaces
+    """)
+
+
+def test_igmp_group_source_limit():
+    "Test IGMP source limits."
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    tgen.gears["r1"].vtysh_cmd("""
+        configure terminal
+        interface r1-eth2
+         ip igmp max-sources 4
+        exit
+    """)
+
+    app_helper.run("h1", ["--source=192.168.100.10", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.11", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.12", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.13", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.14", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.15", "232.0.101.10", "h1-eth0"])
+    app_helper.run("h1", ["--source=192.168.100.16", "232.0.101.10", "h1-eth0"])
+
+    topotest.sleep(10, "Waiting to learn IGMP group sources")
+
+    igmp_sources = tgen.gears["r1"].vtysh_cmd("show ip igmp sources json", isjson=True)
+    assert len(igmp_sources["r1-eth2"]["232.0.101.10"]["sources"]) == 4
```

Comment:
Let us avoid doing sleep. Just do a retry and timeout until you get the result you want or timeout.